Feature/carma wm

msmcconnell opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description

Initial implementation of the carma_wm library for working with road networks (ie. lanelet2). 
This PR closely matches the original with some design tweaks based on the lanelet2 api. 

These are

- In the WorldModel classes Downtrack/crosstrack functions have been combined into 3 functions which return a TrackPos stuct. This contains the downtrack/crosstrack values and is easier to implement using existing lanelet2 functionality. 
- In the WMListener classes the getMutex function has been replaced with a getLock(bool) function that returns an std::unique_lock object which I believe will provide better user exception safety and simplicity of use. 
- The getLocalCurvatures function has been simplified to only calculate curvatures for every set of 3 points. To calculate an average curvature would require performing a polynomial fit which would take a larger amount of time to implement and might not provide major benefits. 

## Related Issue

#441 

## Motivation and Context

Implement world model interface design

## How Has This Been Tested?

Unit tests
## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

msmcconnell commented:
The check is passing because I am still working on the unit tests and haven't update the sonar.properties file. Do not merge yet

Supraluminal commented on carma_wm/include/carma_wm/WorldModel.h:
```diff
@@ -0,0 +1,282 @@
+      // Overload == and != operators
+      bool operator==(const TrackPos &other) const {
+        return this == &other 
+          || (this->downtrack == other.downtrack
```
Does this really make sense as a raw double comparison? Should we consider adding an epsilon factor here to ensure some sanity?

msmcconnell replied:
I believe in this case it should be a direct comparison for ==. This way a copy constructed object will evaluate as equal but a computed one will need the user to account for epsilons themselves and therefore be aware of the desired range.

Supraluminal commented on carma_wm/src/WMListenerWorker.cpp:
```diff
@@ -0,0 +1,61 @@
+    world_model_->setMap(new_map);
+
+    // Call user defined map callback
+    if (map_callback_) {
```
Does an empty `std::function` object evaluate to false? Not sure what the initialization behaviors are for it, but it shouldn't be nullptr considering it's a class member, and would thus have it's default constructor invoked.

msmcconnell replied:
It does

Supraluminal commented on carma_wm/src/WMListenerWorker.cpp:
```diff
@@ -0,0 +1,61 @@
+    // TODO Implement when route message has been defined
+    // world_model_->setRoute(route_obj);
+    // Call route_callback_;
+    if (route_callback_) {
```
Same as above.

Supraluminal commented on carma_wm/include/carma_wm/WMListener.h:
```diff
@@ -0,0 +1,106 @@
+#pragma once
```
Do we have a preference for `#pragma once` vs `#ifndef __...___` guards?

msmcconnell replied:
Shuwei and I have been using ```#pragma once``` but I don't have a big preference either way. Technically if we tries to compile this for something other than ubuntu it could cause issues if the compiler did not support pragma, but that seems unlikely at the moment.

msmcconnell commented on carma_wm/README.md:
```diff
@@ -0,0 +1 @@
+TODO this readme and lanelet2 license notice
```
Plan to after unit tests are completed

msmcconnell commented on carma_wm/package.xml:
```diff
@@ -0,0 +1,24 @@
+  <name>carma_wm</name>
+  <version>1.0.0</version>
+  <description>The carma_wm package</description>
+  <maintainer email="carma@todo.todo">carma</maintainer>
```
Will do

msmcconnell commented on carma_wm/src/CARMAWorldModel.cpp:
```diff
@@ -0,0 +1,333 @@
+  void CARMAWorldModel::setMap(lanelet::LaneletMapPtr map) {
+    semantic_map_ = map;
+    // Build routing graph from map
+    lanelet::traffic_rules::TrafficRulesUPtr traffic_rules = lanelet::traffic_rules::TrafficRulesFactory::create(lanelet::Locations::Germany, lanelet::Participants::VehicleCar);
```
It is the only one that currently exists, and I think it will work for our current use case. I will look into making a user story to flesh out a US one.

msmcconnell commented on carma_wm/src/CARMAWorldModel.cpp:
```diff
@@ -0,0 +1,333 @@
+ 
+  CARMAWorldModel::~CARMAWorldModel() {}
+
+  std::tuple<TrackPos, TrackPos> CARMAWorldModel::routeTrackPos(lanelet::ConstArea& area) const {
```
Hmm seems you are right. If I can find an easy way to auto fix it I will. Otherwise I'll leave it as is (since it is everywhere...)

msmcconnell commented:
@Supraluminal I have added unit tests to cover my changes and made the following adjustments based on your comments. 

- Added README content. Package.xml uses carma@dot.gov content
- Applied the ROS C++ clang-format to all source code. 
- After discovering problematic edge cases, I reworked the point to segment matching logic in the following way. 
  - Created matchSegment function which is an expensive but more robust algorithm for matching a point to a segment on a linestring. 
  - Since the matchSegment function is too slow to be used for the whole route I created the IndexedDistanceMap class which is used on route load to pre-compute distances along the route. This allows the routeTrackPos functions to be far more efficient. 
- I have also removed the getAreasBetween function as I discovered it is unlikely to be of much use as Areas in lanelet to are mostly for parking spots and similar non-road space. It had extremely similar logic to getLaneletsBetween so we can add it back in easily if needed in the future. 

I have squashed the previous set of WIP commits given the fairly substantial changes I had to make.

Supraluminal commented on carma_wm/src/CARMAWorldModel.cpp:
```diff
@@ -0,0 +1,644 @@
+  return lanelet::LineString3d(lanelet::utils::getId(), coppied_points);
+}
+
+void CARMAWorldModel::computeDowntrackReferenceLine()
```
Just noting our discussion about the discontinuity in crosstrack whenever a lane change happens along the reference line, worth thinking of ways we can keep that in mind but I think your documentation is sufficient for now.

Supraluminal commented on carma_wm/src/CARMAWorldModel.h:
```diff
@@ -0,0 +1,150 @@
+
+  TrackPos routeTrackPos(const lanelet::ConstLanelet& lanelet) const override;
+
+  TrackPos routeTrackPos(const lanelet::BasicPoint2d& point) const override;
```
Maybe add a note in the doc comment here about the potential for a discontinuous crosstrack

msmcconnell replied:
Will do

Supraluminal commented on carma_wm/src/CARMAWorldModel.cpp:
```diff
@@ -0,0 +1,644 @@
+                                                                                              // variant
+}
+
+// NOTE: See header file for details on source of logic in this function
```
I might just have missed it but I don't see the note in the header file that's mentionedh ere.

msmcconnell replied:
It's in the WorldModel.h file but I can see how this note is unclear. I'll make it more explicit.

msmcconnell commented on carma_wm/src/IndexedDistanceMap.h:
```diff
@@ -0,0 +1,195 @@
+   *
+   * \param ls The linestring to add
+   */
+  void pushBack(const lanelet::LineString2d& ls)
```
Given the minimal functionality in 90% of the functions I figured a header file was easier since it was internal to the library. I can change it if you want.

Supraluminal replied:
Since it's internal, it's probably fine. Was just curious what the reason was.

msmcconnell commented on carma_wm/src/CARMAWorldModel.cpp:
```diff
@@ -0,0 +1,644 @@
+    else
+    {
+      // If downtrack is negative then iterate over route segments to find preceeding segment
+      const size_t prev_ls_i = ls_i - 1;
```
prev_ls_i will never be negative because of the ls_i == 0 check above.

Supraluminal replied:
Ah, I follow now. I missed that in my read through but that seems to cover the case I'm thinking of.

msmcconnell commented on carma_wm/src/CARMAWorldModel.cpp:
```diff
@@ -0,0 +1,644 @@
+    else
+    {
+      // If downtrack is greater then seg length then we need to find the succeeding segment
+      auto next_centerline = lanelet::utils::to2D(shortest_path_centerlines_[ls_i + 1]);  // Get prev centerline
```
ls_i == shortest_path_centerlines_.size() - 1 check above will prevent that
